luna/gateware/usb/devices/hid.py

```diff
@@ -0,0 +1,123 @@
+from nmigen                                   import Elaboratable, Module, Signal, Cat
+
+from ...stream                                import StreamInterface
+from ..usb2.device                            import USBDevice
+from ..usb2.request                           import USBRequestHandler, StallOnlyRequestHandler
+from ..usb2.endpoints.status                  import USBSignalInEndpoint
+
+from usb_protocol.types                       import USBRequestType,USBTransferType
+from usb_protocol.emitters                    import DeviceDescriptorCollection
+from usb_protocol.emitters.descriptors.hid    import HIDDescriptor, HIDPrefixes
+
+
+
+
+class HIDDevice(Elaboratable):
+
+    _STATUS_ENDPOINT_NUMBER = 1
+
+    def __init__(self, *, bus, idVendor, idProduct,
+        manufacturer_string="LUNA",
+        product_string="HID Register Reader",
+        serial_number=None, max_packet_size=64,
+        usage_page=0x01, usage=0x0):
+
+        self._bus                 = bus
+        self._idVendor            = idVendor
+        self._idProduct           = idProduct
+        self._manufacturer_string = manufacturer_string
+        self._product_string      = product_string
+        self._serial_number       = serial_number
+        self._max_packet_size     = max_packet_size
+        self._usage_page          = usage_page
+        self._usage               = usage
+
+        #
+        # I/O port
+        #
+        self.connect = Signal()
+        self.inputs = []
+
+    def _int_to_le_bytes(self, i):
+        for n in (1, 2, 4):
+            try:
+                return int.to_bytes(i, n, byteorder="little")
+            except OverflowError:
+                pass
+        raise OverflowError("Value cannot be represented in 4 bytes")
+
+    def add_input(self, signal, input_range=None, usage=0x0):
+        if(input_range == None):
+            shape = signal.shape()
+            if(shape.signed):
+                input_range = range(-(1 << (shape.width - 1)) + 1, 1 << (shape.width - 1))
+            else:
+                input_range = range(0, (1 << shape.width) - 1)
+        self.inputs.append((signal, input_range, usage))
+
+
+    def create_descriptors(self):
```

Comment:
This is a very opinionated and strict way of handling of handling the report descriptors. HID descriptors are very different from the other USB descriptors. They are extremely dynamic, HID packets have no standard format, the format is given by the report descriptor.

IMO it would be much better to receive the report descriptor in `__init__` and not have any opinionated declaration like this. If you want, you could then subclass `HIDDevice` and define a few useful devices.

Looking at this report descriptor declaration, it seems to me like you are using it to create device with the vendor page (am I right?), in which case, you could create a `VendorHIDDevice` with almost the exact same code :blush:.

Reply:
I'm actually using the generic desktop controls page with no usage. This causes Linux to create a generic HID input item, and allows users to easily interact with the resulting HID device. My goal here is to create a device that an end user can implement without any knowledge of USB HID, and have it work consistently (ish, Windows is a bit more picky about report formatting). That being said, I'd be plenty happy to change class names or accept changes that allow for more flexibility without requiring HID knowledge to get a device working.

Reply:
I do suppose an option is to auto-generate a working report descriptor if the descriptor in __init__ is None?

Reply:
Yes, that would be great!

Generating a HID report descriptor is a very specific to what you are actually trying to do. Hence my suggestion to subclass `HIDDevice`. The idea would be to subclass into classes that could generate a correct descriptor for common tasks. This would allow users without HID knowledge to create devices. Does that make sense?

I do not really follow your use-case. What is your goal with it? Just to communicate with the device or to report data? If you just want to communicate, a better way would be to use the vendor usage page, which is designed for this. The kernel will still export a hidraw device node, which you can interact with.

Reply:
Re: use cases, I'm currently using python-evdev to poll out register values using this gateware (see https://github.com/TiltMeSenpai/orangecrab_sketches/blob/dev/orangecrab_sketches/usb_hostside.py). This requires Linux to export a /dev/input/event* item, which (as far as I can tell) gets created by the generic HID driver, hence the generic desktop controls page.

Re: HIDDevice subclassing, I think about the only thing a pregenerated report descriptor allows you to do is to use String Index report items to name your input/output fields. Pretty much anything else (such as changing the Collection type) will change the report format, and at that point, I'm not super convinced that a `USBSignalInEndpoint` will still be the right solution. If we let users define their own interrupt endpoints for input/output, the only thing we're really doing is generating descriptors. Is that worth its own class at that point?

Reply:
Well, this is very Linux specific and slightly outside the HID spec. You are using the Generic Desktop usage page with the usage 0, which is undefined.

Generating the report descriptor would be helpful for people wanting to export devices such as a mice, keyboards, joysticks, FIDO keys, vendor protocols, etc. We can generate the report descriptor and give them a way to construct the report by just plugging the field values. For your case, you could create a `GenericInputHIDDevice`.

I still maintain my position on this being a very specific implementation targeted to your use case, and not something I would want in a generic class such as this.

If you do not want a subclass, you could just create a method to generate the report descriptor, but I think subclassing would solve the problem in a nicer way.

Reply:
Hmm... Could you not create a mouse/keyboard/joystick/vendor-defined with the existing implementation? (I think FIDO requires a computer -> LUNA endpoint, which I currently don't have implemented). The only thing the current implementation actually prescribes is all of the report data is consolidated under a single report ID in an Application collection. You can define global Usage Page/Usage and per-item Usages as is, they just default a Generic Desktop page with all Undefined usages. If we move outside of use cases where a `USBSignalInEndpoint` is appropriate, the only thing the generic HID class will do is generate/place the HID class/HID report descriptors, which feels like `python-usb-protocol`'s job. If you feel that this is still appropriate, I'm happy to make the change, but I'm hesitant to blur the package purposes like this.

Reply:
> Hmm... Could you not create a mouse/keyboard/joystick/vendor-defined with the existing implementation?

No, because you are locked to 1 single usage. All you can do is create a device that report a X axis for eg. not a device that export an X an Y axis. You can of course fix this, but if you start trying to fix different use cases things will get out of hand very quickly. You are creating a generalized report descriptor constructor

> Hmm... Could you not create a mouse/keyboard/joystick/vendor-defined with the existing implementation? (I think FIDO requires a computer -> LUNA endpoint, which I currently don't have implemented).

Yes, this is an example. But I assume some day this is something we  will want to support, at least I have interest in it.

> The only thing the current implementation actually prescribes is all of the report data is consolidated under a single report ID in an Application collection.

Hum, what do you mean? You are not using any report ID.

> If we move outside of use cases where a `USBSignalInEndpoint` is appropriate, the only thing the generic HID class will do is generate/place the HID class/HID report descriptors, which feels like `python-usb-protocol`'s job. If you feel that this is still appropriate, I'm happy to make the change, but I'm hesitant to blur the package purposes like this.

Perhaps, this is flexible, I just don't think the current implementation should have such opinionated report descriptor generator.

Anyway, I think is a question for the maintainers. They are the ones giving support for the software (I am available to comment and help out if they want) at the end of the day... If they think the current approach is fine, then go ahead! I just think if things are done a bit differently, it might avoid some long term pain in the future. When you introduce an API such as this, you are inevitable locked into it, breaking changes are something you want to avoid.